### PR TITLE
feat: support EVM memos / UTXOs OP_RETURNs as memos

### PR DESCRIPTION
**Pull request by gomesalexandre**
### Description

- support memos for EVM chains as first-class citizens of `BuildTxInput`, since it was not previously possible to pass arbitrary calldata
- support memos across chains as a first-class citizens of `GetFeeDataInput`, so we can `estimateFees()` in web 

See web consumption in https://github.com/shapeshift/web/pull/3697/files

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1180** <a href="https://app.graphite.dev/github/pr/shapeshift/lib/1180" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/lib/1180?utm_source=stack-comment).

File: packages/chain-adapters/src/types.ts [hunk 1]
```diff
@@ -228,6 +228,9 @@ type ChainSpecificGetFeeDataInput<T> = ChainSpecific<
   }
 >
 export type GetFeeDataInput<T extends ChainId> = {
+  // Optional hex-encoded calldata
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/types.ts:231]:
```suggestion
  // Optional hex-encoded calldata for EVM chains, UTF-8 for others
```

File: packages/chain-adapters/src/evm/EvmBaseAdapter.ts [hunk 2]
```diff
@@ -238,7 +239,7 @@ export abstract class EvmBaseAdapter<T extends EvmChainId> implements IChainAdap
       value = erc20Balance
     }
 
-    const data = contractData ?? (await getErc20Data(to, value, contractAddress))
+    const data = memo || contractData || (await getErc20Data(to, value, contractAddress))
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/evm/EvmBaseAdapter.ts:242]: Safer, memo should be undefined or a non-empty string but you never know what consumers might pass

File: packages/chain-adapters/src/evm/EvmBaseAdapter.ts [hunk 3]
```diff
@@ -187,7 +187,7 @@ export abstract class EvmBaseAdapter<T extends EvmChainId> implements IChainAdap
           tx.value = bnOrZero(account.balance).minus(fee).toString()
         }
       }
-      const data = await getErc20Data(to, tx.value, erc20ContractAddress)
+      const data = tx.memo || (await getErc20Data(to, tx.value, erc20ContractAddress))
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/evm/EvmBaseAdapter.ts:190]: Safer than `??` WRT nullish strings, memo should be undefined or a non-empty string but you never know what consumers might pass

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**